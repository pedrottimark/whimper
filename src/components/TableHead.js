/* @flow */

import React, {Component} from 'react';

import type {
  FieldKey,
  Fields,
  FilterRecords,
  SortRecords,
  SortingCriterion,
  View,
} from '../types';

type Props = {|
  addRow: Function,
  count: number,
  fields: Fields,
  filterRecords: FilterRecords,
  sortRecords: SortRecords,
  view: View,
|};

export const ascending = '\u25B2'; // black up-pointing triangle
export const descending = '\u25BC'; // black down-pointing triangle

function sortingIndicator(fieldKey: FieldKey, sortingCriterion?: SortingCriterion) {
  let title = '';
  let text;
  if (sortingCriterion && sortingCriterion.fieldKey === fieldKey) {
    if (sortingCriterion.descending) {
      title = 'descending';
      text = descending;
    } else {
      title = 'ascending';
      text = ascending;
    }
  }

  return {
    title,
    text,
  };
}

class TableHead extends Component {
  props: Props;

  _addRow = () => {
    this.props.addRow();
  }

  _filterRecords = (e: Event) => {
    const target = ((e.target: any): HTMLInputElement);
    this.props.filterRecords(target.value);
  }

  _resortRows = () => {
    this.props.sortRecords();
  }

  render() {
    const {count, fields, sortRecords, view} = this.props;
    const addRow = '\u2795'; // HEAVY PLUS SIGN
    const placeholder = 'ðŸ”Ž filter rows'; // 1F50E RIGHT-POINTING MAGNIFYING GLASS
    const sortingCriterion = view.sorting[0];
    return (
      <thead>
        <tr key="interaction">
          <th key="" scope="col" title="add row" onClick={this._addRow}>
            <button>{addRow}</button>
          </th>
          {
            fields.length !== 0 && (
              <th key="colgroup" scope="colgroup" colSpan={fields.length}>
                <input placeholder={placeholder} onChange={this._filterRecords} />
              </th>
            )
          }
        </tr>
        <tr key="headings">
          <th key="" title="reset sort order" onClick={this._resortRows} scope="col">{count}</th>
          {
            fields.map(({key, label}) => {
              const {title, text} = sortingIndicator(key, sortingCriterion);
              return (
                <th key={key} onClick={() => { sortRecords(key); }} scope="col">
                  <span>{label}</span>
                  <abbr title={title}>{text}</abbr>
                </th>
              );
            })
          }
        </tr>
      </thead>
    );
  }
}

export default TableHead;
