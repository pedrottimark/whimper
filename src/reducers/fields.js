/* @flow */

import type {
  Action,
  Field,
  Fields,
  FieldValue,
} from '../types';

// production app might get fields from external source :)
export const fieldsReceived: Fields = [
  {
    key: 'when',
    label: 'when',
    type: 'number',
    defaultValue: new Date().getFullYear(),
  },
  {
    key: 'what',
    label: 'what',
    type: 'text',
    defaultValue: '',
  },
  {
    key: 'weep',
    label: 'weep',
    type: 'text',
    defaultValue: '',
  },
];

// Given a field and text input value,
// return a data value to update the property in a record.
export const fieldValue = ({type}: Field, text: string): FieldValue => type === 'number'
  ? parseInt(text, 10)
  : text;

export const fieldsInitial: Fields = [];

export default function (fields: Fields = fieldsInitial, action: Action): Fields {
  switch (action.type) {

  case 'RECEIVE_DATA':
    return action.fields || fields;

  default:
    return fields;

  }
}
