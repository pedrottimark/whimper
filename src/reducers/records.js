/* @flow */

import type {
  Action,
  DeleteRecordAction,
  UpdateFieldAction,
  Fields,
  Record,
  RecordId,
  Records,
} from '../types';

/*
// Convert internal collection from and to external JSON.
export const recordsFromJSON = (array: Record[]): Records => array;
export const recordsToJSON = (records: Records): Record[] => records;
*/

// Initial state of records before they have been received from external source.
export const recordsInitial: Records = [];

// A record object has an implicit and temporary id: its index in the array.
export default function (records: Records = recordsInitial, action: Action): Records {
  switch (action.type) {

  case 'CREATE_RECORD':
    // Create new record as first element of the collection.
    return [action.record].concat(records);

  case 'RECEIVE_DATA':
    // Replace initial records, except when whimper starts for the first time.
    return action.records || records;

  case 'UPDATE_FIELD': {
    // Destructuring and type cast to avoid warnings in Flow 0.35.0
    const {recordId, fieldKey, value} = (action: UpdateFieldAction);
    return records.map(record => record.id === recordId
      ? Object.assign({}, record, {[fieldKey]: value})
      : record
    );
  }

  case 'DELETE_RECORD': {
    // Destructuring and type cast to avoid warnings in Flow 0.35.0
    const {recordId} = (action: DeleteRecordAction);
    return records.filter(record => record.id !== recordId);
  }

  default:
    return records;

  }
}

// A very naive algorithm, instead of hashing ;)
const unusedId = (records: Records): RecordId =>
  records.reduce((id, record) => id <= record.id ? record.id + 1 : id, 0);

// Given fields, return a record with default values.
// For example, to create a new record.
export const recordDefault = (fields: Fields, records: Records): Record =>
  fields.reduce((record, {key, defaultValue}) => {
    record[key] = defaultValue;
    return record;
  }, {
    id: unusedId(records)
  });
