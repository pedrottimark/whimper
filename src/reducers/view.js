import type {
  Action,
  FieldValue,
  Fields,
  FilteringCriterion,
  Record,
  View,
} from '../types';

const filteringInitial = '';
const sortingInitial = [];
export const viewInitial: View = {
  filtering: filteringInitial,
  sorting: sortingInitial,
};

export default function (view: View = viewInitial, action: Action): View {
  switch (action.type) {

    case 'SEARCH_RECORDS':
      return Object.assign({}, view, {
        filtering: action.filtering
      });

    case 'SORT_RECORDS':
      // Add current sorting criteria as first element of array.
      // Toggle descending if previous first element has same fieldKey.
      // The array can have at most one element for any fieldKey.
      return Object.assign({}, view, {
        sorting: action.fieldKey
          ? [{
              fieldKey: action.fieldKey,
              descending: view.sorting.length !== 0 &&
                view.sorting[0].fieldKey === action.fieldKey &&
                !view.sorting[0].descending,
            }].concat(view.sorting.filter(({fieldKey}) => fieldKey !== action.fieldKey))
          : sortingInitial
      });

    default:
      return view;
  }
}

// Return sorting comparision of field values.
const compareFieldValues = (fieldValueA: FieldValue, fieldValueB: FieldValue, descending: boolean): number => {
  const comparison = typeof fieldValueA === 'number' && typeof fieldValueB === 'number'
    ? fieldValueA - fieldValueB
    : String(fieldValueA).localeCompare(String(fieldValueB));

  return descending ? -comparison : comparison;
};

// Return sorting comparision of records for zero or more fields.
// Assume that the sorting is stable (for example, OrderedMap immutable collection).
const compareRecords = (recordA: Record, recordB: Record, sorting: SortingCriterion[]) : number =>
  sorting.reduce((comparison, {fieldKey, descending}) =>
    comparison || // compare field values only if all preceding comparisons are zero
    compareFieldValues(recordA[fieldKey], recordB[fieldKey], descending),
  0);

// Return whether the record has at least one field which includes the filtering string.
const filterRecord = (record: Record, fields: Fields, filtering: FilteringCriterion): boolean =>
  filtering.length === 0 || // not filtering
  fields.some(({key: fieldKey}) => record[fieldKey].toString().toLowerCase().includes(filtering));

// Return records to display in a view, given the sorting and filtering critera.
// A record id is its index in the currently stored list of records.
export const recordsInView = (records: Records, fields: Fields, {filtering, sorting}: View): Records =>
  records
    .filter((record) => filterRecord(record, fields, filtering))
    .sort((recordA, recordB) => compareRecords(recordA, recordB, sorting));
