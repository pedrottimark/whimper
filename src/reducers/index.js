import { combineReducers } from 'redux';

// child reducers
import fields from './fields';
import records from './records';
import view from './view';

// root reducer
export default combineReducers({
  fields,
  records,
  view,
});
