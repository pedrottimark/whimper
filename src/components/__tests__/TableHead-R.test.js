/* @flow */

import React from 'react';
import renderer from 'react-test-renderer';
import {
  irrelevant,
  relevantTestObject,
  renderAsTestObject,
} from '../../testing/relevant-test-object';

import {
  sortRecords,
} from '../../actions';
import {
  fieldsInitial,
  fieldsReceived as fields,
} from '../../reducers/fields';
import {viewInitial} from '../../reducers/view';

import TableHead, {
  ascending,
  descending,
} from '../TableHead';

const addRow = () => {};

describe('TableHead', () => {
  it('renders button, count, but no fields', () => {
    expect(renderer.create(
      <TableHead
        addRow={addRow}
        count={0}
        fields={fieldsInitial}
        sortRecords={sortRecords}
        view={viewInitial}
      />
    ).toJSON()).toMatchSnapshot();
  });

  it('renders button, count, and fields', () => {
    expect(renderer.create(
      <TableHead
        addRow={addRow}
        count={7}
        fields={fields}
        sortRecords={sortRecords}
        view={viewInitial}
      />
    ).toJSON()).toMatchSnapshot();
  });

  it('renders ascending indicator in `what` heading', () => {
    expect(renderAsTestObject(
      <TableHead
        addRow={addRow}
        count={7}
        fields={fields}
        sortRecords={sortRecords}
        view={Object.assign({}, viewInitial, {
          sorting: [
            {
              fieldKey: 'what',
              descending: false,
            },
          ]
        })}
      />
    ).children[1]).toMatchObject(relevantTestObject(
      <tr>
        <th>{irrelevant}</th>
        <th>
          <span>{irrelevant}</span>
          <abbr title=""></abbr>
        </th>
        <th>
          <span>{irrelevant}</span>
          <abbr title="ascending">{ascending}</abbr>
        </th>
        <th>
          <span>{irrelevant}</span>
          <abbr title=""></abbr>
        </th>
      </tr>
    ));
  });

  it('renders descending indicator in `when` heading', () => {
    expect(renderAsTestObject(
      <TableHead
        addRow={addRow}
        count={7}
        fields={fields}
        sortRecords={sortRecords}
        view={Object.assign({}, viewInitial, {
          sorting: [
            {
              fieldKey: 'when',
              descending: true,
            },
            {
              fieldKey: 'what',
              descending: false,
            },
          ]
        })}
      />
    ).children[1]).toMatchObject(relevantTestObject(
      <tr>
        <th>{irrelevant}</th>
        <th>
          <span>{irrelevant}</span>
          <abbr title="descending">{descending}</abbr>
        </th>
        <th>
          <span>{irrelevant}</span>
          <abbr title=""></abbr>
        </th>
        <th>
          <span>{irrelevant}</span>
          <abbr title=""></abbr>
        </th>
      </tr>
    ));
  });
});
