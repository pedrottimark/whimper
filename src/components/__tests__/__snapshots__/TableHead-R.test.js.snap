// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableHead renders button, count, and fields 1`] = `
<thead>
  <tr>
    <th
      onClick={[Function]}
      scope="col"
      title="add row"
    >
      <button>
        âž•
      </button>
    </th>
    <th
      colSpan={3}
      scope="colgroup"
    >
      <input
        onChange={[Function]}
        placeholder="ðŸ”Ž filter rows"
      />
    </th>
  </tr>
  <tr>
    <th
      onClick={[Function]}
      scope="col"
      title="reset sort order"
    >
      7
    </th>
    <th
      onClick={[Function]}
      scope="col"
    >
      <span>
        when
      </span>
      <abbr
        title=""
      >
        
      </abbr>
    </th>
    <th
      onClick={[Function]}
      scope="col"
    >
      <span>
        what
      </span>
      <abbr
        title=""
      >
        
      </abbr>
    </th>
    <th
      onClick={[Function]}
      scope="col"
    >
      <span>
        weep
      </span>
      <abbr
        title=""
      >
        
      </abbr>
    </th>
  </tr>
</thead>
`;

exports[`TableHead renders button, count, but no fields 1`] = `
<thead>
  <tr>
    <th
      onClick={[Function]}
      scope="col"
      title="add row"
    >
      <button>
        âž•
      </button>
    </th>
  </tr>
  <tr>
    <th
      onClick={[Function]}
      scope="col"
      title="reset sort order"
    >
      0
    </th>
  </tr>
</thead>
`;
